java/ql/src/experimental/Security/CWE/CWE-094/JythonInjection.ql

```diff
@@ -0,0 +1,123 @@
+class InterpretExprMethod extends Method {
+  InterpretExprMethod() {
+    this.getDeclaringType().getAnAncestor*() instanceof PythonInterpreter and
+    (
+      getName().matches("exec%") or
+      hasName("eval") or
+      hasName("compile") or
+      getName().matches("run%")
+    )
+  }
+}
+
+/** The class `org.python.core.BytecodeLoader`. */
+class BytecodeLoader extends RefType {
+  BytecodeLoader() { this.hasQualifiedName("org.python.core", "BytecodeLoader") }
+}
+
+/** Holds if a Jython expression if evaluated, compiled or executed. */
+predicate runCode(MethodAccess ma, Expr sink) {
+  exists(Method m | m = ma.getMethod() |
+    m instanceof InterpretExprMethod and
+    sink = ma.getArgument(0)
+  )
+}
+
+/** A method that loads Java class data. */
+class LoadClassMethod extends Method {
+  LoadClassMethod() {
+    this.getDeclaringType().getAnAncestor*() instanceof BytecodeLoader and
+    (
+      hasName("makeClass") or
+      hasName("makeCode")
+    )
+  }
+}
+
+/** Holds if a Java class file is loaded. */
+predicate loadClass(MethodAccess ma, Expr sink) {
+  exists(Method m, int i | m = ma.getMethod() |
+    m instanceof LoadClassMethod and
+    m.getParameter(i).getType() instanceof Array and // makeClass(java.lang.String name, byte[] data, ...)
+    sink = ma.getArgument(i)
+  )
+}
+
+/** The class `org.python.core.Py`. */
+class Py extends RefType {
+  Py() { this.hasQualifiedName("org.python.core", "Py") }
+}
+
+/** A method that compiles code with `Py`. */
+class PyCompileMethod extends Method {
+  PyCompileMethod() {
+    this.getDeclaringType().getAnAncestor*() instanceof Py and
+    getName().matches("compile%")
+  }
+}
+
+/** Holds if source code is compiled with `PyCompileMethod`. */
+predicate compile(MethodAccess ma, Expr sink) {
+  exists(Method m | m = ma.getMethod() |
+    m instanceof PyCompileMethod and
+    sink = ma.getArgument(0)
+  )
+}
+
+/** Sink of an expression loaded by Jython. */
+class CodeInjectionSink extends DataFlow::ExprNode {
+  CodeInjectionSink() {
+    runCode(_, this.getExpr()) or
+    loadClass(_, this.getExpr()) or
+    compile(_, this.getExpr())
+  }
+
+  MethodAccess getMethodAccess() {
+    runCode(result, this.getExpr()) or
+    loadClass(result, this.getExpr()) or
+    compile(result, this.getExpr())
+  }
+}
+
+class CodeInjectionConfiguration extends TaintTracking::Configuration {
+  CodeInjectionConfiguration() { this = "CodeInjectionConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof RemoteFlowSource
+    or
+    source instanceof LocalUserInput
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof CodeInjectionSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    // @RequestBody MyQueryObj query; interpreter.exec(query.getInterpreterCode());
+    exists(MethodAccess ma | ma.getQualifier() = node1.asExpr() and ma = node2.asExpr())
```

Comment:
Definitely need to restrict `ma` more; currently this introduces an additional taint step for every method call.

Reply:
The <code>isAdditionalTaintStep</code> predicate is for an edge case - Python code in a POJO object of remote input source such as <code>Spring RequestBody</code> but the Python code string is not implemented as an instance attribute e.g. in the referenced sample project. Therefore it's not evaluated as a <code>GetterMethod</code> thus not detected by the following CodeQL library: https://github.com/github/codeql/blob/bc8c55836a24317cb06d25b014102364c9799fc5/java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll#L248-L250

I've added the missing constraint to the additional taint step. Or please let me know if you would like to remove the whole <code>isAdditionalTaintStep</code> predicate if that helps to clarify the query.

Reply:
Removed also as per the discussion with @pwntester below